include/unifex/async_scope.hpp

```diff
@@ -655,53 +653,100 @@ class _attached_op<Sender, Receiver>::type final {
   ~type() {
     auto scope = scope_ref();
     if (scope) {
-      unifex::deactivate_union_member(op_);
+      op_.destruct();
       // started => receiver responsible for cleanup
-      if (!has_started()) {
+      if (incref() == 1) {
         record_done(scope);
       }
-    } else {
-      unifex::deactivate_union_member(receiver_);
     }
   }
 
-  auto get_token() { return stopSource_.get_token(); }
+  void request_stop() noexcept {
+    // increment from 0 means lost race with a completion method so
+    // no-op and let the in-flight completion complete
+    if (incref() == 0) {
+      return;
+    }
+    stopSource_.request_stop();
+    deliver_result([](auto&& r) noexcept { unifex::set_done(std::move(r)); });
+  }
+
+  // reduce size of the templated `deliver_result`
+  async_scope* deliver_result_prelude() noexcept {
+    if (decref() != 1) {
+      return nullptr;
+    }
+    auto scope = scope_ref();
+    UNIFEX_ASSERT(scope != nullptr);
+    deregister_callbacks();
+    return scope;
```

Comment:
This can be reduced to one atomic op, too.
```suggestion
    auto scope = scope_.fetch_sub(1u, std::memory_order_acq_rel);
    if ((scope & ~mask) != 1) {
      UNIFEX_ASSERT((scope & ~mask) == 2);
      return nullptr;
    }
    UNIFEX_ASSERT((scope & mask) != 0u);
    deregister_callbacks();
    return reinterpret_cast<async_scope*>(scope & mask);
```